content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -1,38 +1,408 @@
+the language evolves, just in case somebody does use them. 
 
 
 Querying Representation Choices
 -------------------------------
 
-.. todo::
+Earlier we mentioned the ability to specify a record type's layout, or a
+variable's address in memory. These are just some of the possibilities
+supported by the language in section 13. For any representation
+characteristic you can specify, you can also query that characteristic.
+That can be very handy, especially if you did not specify it and want to
+know what the compiler chose for the representation.
 
-    Complete section!
+For example, say you want to express an "overlay," in which an object of
+one type is placed at the same memory location as a distinct object of a
+distinct type, thus overlaying one object over the other. Doing so
+allows you to interact with those memory locations in more than one way:
+one way per type, specifically. This is known as "type punning" in
+computer programming. There are other was to achieve that effect as
+well, but realize that doing so circumvents the static strong typing
+used by Ada to protect us from ourselves and from others. Use it with
+care!
+
+For a concrete example, imagine a sensor providing a stream of bytes as
+input, and you know that groups of those bytes comprise a composite
+value, i.e., a value of some record type. The buffered array of inputs
+bytes would be declared in terms of a numeric, probably unsigned integer
+Byte type, whereas the composite type would, of course, be the record
+type. The clients of the sensor driver call a driver-defined procedure
+to get the latest value, passing an output argument of the record type. The
+procedure gets the next buffered bytes, converts them to the record type
+and puts the value in the parameter, maybe checks it for validity and so
+on, and then the procedure returns. An overlay would make that simple.
+We would declare an object of the record type located at the same
+address in the buffer as the starting byte representing the next
+composite value.
+
+But there's a problem: inside the procedure we want to overlay our 
+record object onto the address of one of the bytes in the buffer. But we 
+didn't declare the Buffer ourselves, at a known location. It is passed 
+to the procedure as an argument. Therefore, when writing the procedure 
+we don't already know the address of the buffer, nor the addresses of 
+the bytes within. As a result, we don't know the address to use in our 
+overlay. 
+
+Of course, there is a solution. We can query the address of objects, and
+other things too, but objects, especially variables, are the most common
+case. In particular, we can say `X'Address` to query the (starting)
+address of object X, including the case where X is a 
+component of an array. With that information we know what address to
+specify for our record object overlay.
+
+Here's the generic procedure again, now showing the overlay:
+
+.. code-block:: ada
```

Comment:
Similar to my previous comment, this should be converted to a complete source-code example using `.. code::`.

Reply:
I disagree.